### PR TITLE
boot: Try to detect overlapping PE sections

### PR DESCRIPTION
**Pull request by medhefgo**

### CODE REVIEW

File: src/boot/efi/pe.c [hunk 1]
```diff
@@ -151,9 +151,17 @@ static void locate_sections(
         assert(offsets);
         assert(sizes);
 
+        size_t prev_section_addr = 0;
+
         for (UINTN i = 0; i < n_table; i++) {
                 const PeSectionHeader *sect = section_table + i;
 
+                if (in_memory) {
+                        if (prev_section_addr > sect->VirtualAddress)
```
Comment (poettering) [inline src/boot/efi/pe.c:160]: hmm, are sections supposed to be ordered by offset in the file? Id don't think so, do they?

File: src/boot/efi/pe.c [hunk 2]
```diff
@@ -151,9 +151,17 @@ static void locate_sections(
...
+                                log_error_stall(u"Overlapping PE sections detected. Boot may fail due to image memory corruption!");
+                        prev_section_addr = sect->VirtualAddress + sect->VirtualSize;
```
Comment (poettering) [inline src/boot/efi/pe.c:162]: maybe if we are at it also do an overflow check?

File: src/boot/efi/pe.c (hunk 1 above)
Comment (medhefgo) [reply src/boot/efi/pe.c:160]: As far as I can tell, sections are sorted by VMA.

File: src/boot/efi/pe.c (hunk 2 above)
Comment (medhefgo) [reply src/boot/efi/pe.c:162]: The image is already in memory, relocated and running. There's hardly any point in doing that now.

File: src/boot/efi/pe.c (hunk 1 above)
Comment (poettering) [reply src/boot/efi/pe.c:160]: indeed, spec says so. 

https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#section-table-section-headers

Maybe add a comment here, saying sections must be ordered by VirtualAddress, and maybe even add a reference to the spec.

Hmm, maybe also check this order while we iterate through the thing, given such a check is basically free.

File: src/boot/efi/pe.c (hunk 2 above)
Comment (poettering) [reply src/boot/efi/pe.c:162]: you never know what quirks the firmware has. As a matter of hygiene I'd always check things like that in new code.

File: src/boot/efi/pe.c (hunk 1 above)
Comment (medhefgo) [reply src/boot/efi/pe.c:160]: And if you looked at the code you'd see that it does exactly that. :facepalm:

File: src/boot/efi/pe.c (hunk 2 above)
Comment (medhefgo) [reply src/boot/efi/pe.c:162]: This has never been an issue so I really fail to see the point in doing so. This was just supposed to be a safety measure to detect badly assembled unified images that are already running and you wanna make me jump through pointles hoops? nope. Use this or stick your head in the sand.

And no, detecting this so the image never even gets shown in the boot menu is a terrible idea. People will complain that systemd-boot is too stupid to pick up its own menu entries.

Comment (poettering) [reply src/boot/efi/pe.c:162]: dunno, it's just a matter of hygiene. For new code we try to be more careful than we used to and always assume that something like a supercharged `-ftrapw` mode (that also covers unsigned ints) was in effect.

Comment (poettering) [reply src/boot/efi/pe.c:162]: (i.e. clang has "intsan" for that as i understand)

**Pull request merged**